compiler/rustc_middle/src/thir.rs

```diff
@@ -695,17 +695,32 @@ impl<'tcx> fmt::Display for Pat<'tcx> {
                 Ok(())
             }
             PatKind::Variant { ref subpatterns, .. } | PatKind::Leaf { ref subpatterns } => {
-                let variant = match self.kind {
-                    PatKind::Variant { adt_def, variant_index, .. } => {
-                        Some(adt_def.variant(variant_index))
-                    }
-                    _ => self.ty.ty_adt_def().and_then(|adt| {
-                        if !adt.is_enum() { Some(adt.non_enum_variant()) } else { None }
+                let variant_and_name = match self.kind {
+                    PatKind::Variant { adt_def, variant_index, .. } => ty::tls::with(|tcx| {
+                        let variant = adt_def.variant(variant_index);
+                        let adt_did = adt_def.did();
+                        let name = if tcx.get_diagnostic_item(sym::Option) == Some(adt_did)
+                            || tcx.get_diagnostic_item(sym::Result) == Some(adt_did)
```

Comment:
This seems like it might give incorrect suggestions when using [`#![no_implicit_prelude]`](https://doc.rust-lang.org/reference/names/preludes.html#the-no_implicit_prelude-attribute).

Reply:
I guess, but that could probably be said about plenty of other suggestions where we're currently render. They're, at most, best effort.

I don't think this needs to be fixed, especially because it's basically impossible to determine what the proper shortest path to refer to an item is in any module, due to `rustc_resolve` not exposing, and it would be unnecessarily verbose to print the full untrimmed path (e.g. `std::option::Option::Some`).

Reply:
cc @estebank I remember some tests getting updated that now figure out the best import path to use, but  I may be totally off here

Reply:
I believe the change you remember @oli-obk was about the ordering of suggestions (it filters `core::` paths if the `std::` is available too), not for this. I think that we eventually would want to rebuild suggestions so that we can ask for paths in a specific DefId or HirId so that it is always accurate for that scope, but we're far from there.

I'm ok with landing this as is.